pydantic_evals/pydantic_evals/assessments/spec.py

```diff
@@ -0,0 +1,234 @@
+            bound_sync_function = cast(SyncBoundAssessmentFunction[InputsT, OutputT, MetadataT], bound_function)
+            return Assessment[InputsT, OutputT, MetadataT](spec=spec, function=bound_sync_function)
+
+    async def execute(self, ctx: ScoringContext[InputsT, OutputT, MetadataT]) -> list[AssessmentDetail]:
+        if inspect.iscoroutinefunction(self.function):
+            async_function = cast(AsyncBoundAssessmentFunction[InputsT, OutputT, MetadataT], self.function)
+            results = await async_function(ctx)
+        else:
+            sync_function = cast(SyncBoundAssessmentFunction[InputsT, OutputT, MetadataT], self.function)
+            results = await anyio.to_thread.run_sync(sync_function, ctx)
+
+        results = _convert_to_mapping(results, scalar_name=self.spec.call)
+
+        details: list[AssessmentDetail] = []
+        for name, result in results.items():
+            if not isinstance(result, AssessmentResult):
+                result = AssessmentResult(value=result)
+            details.append(AssessmentDetail(name=name, value=result.value, reason=result.reason, source=self.spec))
+
+        return details
+
+    @staticmethod
+    def _validate_against_registry(
+        registry: AssessmentRegistry[Any, Any, Any], case_name: str, spec: AssessmentSpec
+    ) -> None:
+        function = registry.get(spec.call)
+        if function is None:
+            raise ValueError(
+                f'Assertion call {spec.call!r} is not registered. Registered choices: {list(registry.keys())}'
+            )
+
+        signature = inspect.signature(function)
+        first_param = list(signature.parameters.values())[0]
+        type_hints = _typing_extra.get_function_type_hints(function)
+        type_hints.pop(first_param.name)
+        type_hints.pop('return', None)
+        for p in signature.parameters.values():
+            if p.default is not p.empty:
+                type_hints[p.name] = NotRequired[type_hints[p.name]]
+
+        td = TypedDict(function.__name__, type_hints)  # type: ignore
+        td.__pydantic_config__ = {'extra': 'forbid'}  # type: ignore
+        adapter = TypeAdapter(td)
+        arguments = spec.arguments
+        if not isinstance(arguments, dict):
+            first_kwarg = '__arg'
+            for kwarg in type_hints:
+                first_kwarg = kwarg
+                break
+            arguments = {first_kwarg: arguments}
+
+        try:
+            adapter.validate_python(arguments)
+        except ValidationError as e:
+            error_message = str(e)
+            error_message = error_message.replace(' for typed-dict', ':')
+            raise ValueError(f'Assessment {spec.call!r} in case {case_name!r} has {error_message}')
+
+
+_DEFAULT_REGISTRY: dict[str, AssessmentFunction[Any, Any, Any]] = {}
+
+F = TypeVar('F', bound=AssessmentFunction[Any, Any, Any])
+
+
+def assessment(f: F) -> F:
+    """Decorator that registers an assessment function in the default registry."""
```

Comment:
Users are meant to use this decorator, right? If so then it should probably be explicitly tested and also documented more clearly, the default registry will mean nothing to users.

Reply:
.. like everything in this PR 😅